### PR TITLE
Qualification tool: Parse expressions in Join execs

### PR DESCRIPTION
**Pull request by nartal1**
This closes https://github.com/NVIDIA/spark-rapids/issues/6064 
This PR adds parser support to the following join execs.
1. BroadcastHashJoin
2. SortMergeJoin
3. BroadcastNestedLoopJoin
4. ShuffledHashJoin

The format in eventlog is almost similar for equijoins. All use same function to parse the exec. 
Added new test for unsupported expressions and updated the current test for supported expressions.

### CODE REVIEW

**Comment by nartal1** [comment]
build

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala [hunk 1]
```diff
@@ -434,6 +434,150 @@ object SQLPlanParser extends Logging {
...
   }
 
+   // This parser is used for BroadcastHashJoin, ShuffledHashJoin and SortMergeJoin
+   def parseEquijoinsExpressions(exprStr: String): (Array[String], Boolean) = {
+     // ShuffledHashJoin [name#11, CEIL(DEPT#12)], [name#28, CEIL(DEPT_ID#27)], Inner, BuildLeft
+     // SortMergeJoin [name#11, CEIL(dept#12)], [name#28, CEIL(dept_id#27)], Inner
+     // BroadcastHashJoin [name#11, CEIL(dept#12)], [name#28, CEIL(dept_id#27)], Inner,
+     // BuildRight, false
+     val parsedExpressions = ArrayBuffer[String]()
+     val pattern = """\[([\w#, +*\\\-\.<>=\`\(\)]+\])""".r
```
Comment (gerashegalov) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala:444]: regex pattern creation typically should be outside method calls to amortize the creation cost.

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala [hunk 2]
```diff
@@ -434,6 +434,150 @@ object SQLPlanParser extends Logging {
...
+     val joinType = joinParams(0).trim
+     // SortMergeJoin doesn't have buildSide, assign empty string in that case
+     val buildSide = if (joinParams.size > 1) {
+       joinParams(1).trim
+     } else {
+       ""
+     }
+     // Get individual expressions which is later used to get the function names.
+     val expressions = joinExprs.split("::").map(_.trim).map(
+       _.replaceAll("""^\[+""", "").replaceAll("""\]+$""", "")).
```
Comment (gerashegalov) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala:459]: if this is called a lot, create regex patterns. 

Probably can do it in a single replaceAll with an OR-ed regex?
```suggestion
       _.replaceAll("""^\[+|\]+$""", "")
```

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala [hunk 3]
```diff
@@ -434,6 +434,150 @@ object SQLPlanParser extends Logging {
...
+       map(_.split(",")).flatten.map(_.trim)
+
+     if (expressions.nonEmpty) {
+       val functionPattern = """(\w+)\(.*\)""".r
+       expressions.foreach { expr =>
+         val functionName = getFunctionName(functionPattern, expr)
+         functionName match {
+           case Some(func) => parsedExpressions += func
+           case _ => // NO OP
+         }
```
Comment (gerashegalov) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala:469]: nit
```suggestion
         functionName.foreach(parsedExpressions += _)
```

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala (hunk 1 above)
Comment (nartal1) [reply tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala]: Thanks for the suggestion. I tend to declare them where it is used but declaring outside the method calls in this case makes sense. Updated in [670c8b1](https://github.com/NVIDIA/spark-rapids/pull/6796/commits/670c8b148bca5ad33e8b230127e787f535ece105)

**Comment by nartal1** [comment]
build

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala [hunk 4]
```diff
@@ -434,6 +434,145 @@ object SQLPlanParser extends Logging {
...
+    // BuildRight, LeftOuter, ((CEIL(cast(id1#1490 as double)) <= cast(id2#1496 as bigint))
+    // AND (cast(id1#1490 as bigint) < CEIL(cast(id2#1496 as double))))
+    val parsedExpressions = ArrayBuffer[String]()
+    // Get joinType and buildSide by splitting the input string.
+    val nestedLoopParameters = exprStr.split(",", 3)
+    val buildSide = nestedLoopParameters(0).trim
+    val joinType = nestedLoopParameters(1).trim
+
+    if (nestedLoopParameters.size > 2) { // condition present on join columns
+      val exprSepAND = if (exprStr.contains("AND")) {
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala:482]: much of this code look the same as the filterExpresions parser, wondering if we can create some common functions between those?

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala [hunk 5]
```diff
@@ -434,6 +434,145 @@ object SQLPlanParser extends Logging {
...
+
+  private def isJoinTypeSupported(joinType: String): Boolean = {
+    joinType match {
+      case JoinType.Cross => true
+      case JoinType.Inner => true
+      case JoinType.LeftSemi => true
+      case JoinType.LeftOuter => true
+      case JoinType.RightOuter => true
+      case JoinType.LeftAnti => true
+      case _ => false
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala:539]: what about FullOuter and ExistenceJoin here?  I thought we supported some of those. Are we not able to tell that here? if not we should add comments

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala (hunk 4 above)
Comment (nartal1) [reply tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala]: created common function for BroadcastNestedLoopJoinExec and FilterExec

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala (hunk 5 above)
Comment (nartal1) [reply tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala:503]: FullOuter is not supported for BroadcastNestedLoopJoin, and for equiJoins there is a caveat that it doesn't support for struct keys. I updated the code to mention it as supported for equiJoins. For ExistenceJoins, added comment that it doesn't show up in event logs.

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala [hunk 6]
```diff
@@ -434,6 +434,113 @@ object SQLPlanParser extends Logging {
...
+    // below condition is met.
+    val buildSideSupported = if (buildSide == BuildSide.BuildLeft) {
+      joinType == JoinType.Inner || joinType == JoinType.Cross ||
+        joinType == JoinType.RightOuter || joinType == JoinType.FullOuter
+    } else if (buildSide == BuildSide.BuildRight) {
+      joinType == JoinType.Inner || joinType == JoinType.Cross ||
+        joinType == JoinType.LeftOuter || joinType == JoinType.LeftSemi ||
+        joinType == JoinType.LeftAnti
+    } else {
+      true
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala:519]: did you mean to have default case true here?

**Review by tgravescs — APPROVED** [review]

**Comment by nartal1** [comment]
build

**Comment by nartal1** [comment]
build

**Pull request merged**